module/behaviour/planning/KickPlanner/src/KickPlanner.cpp

```diff
@@ -83,32 +83,34 @@ namespace behaviour {
                 emit(std::make_unique<WantsToKick>(false));
             });
 
-            on<Trigger<Ball>, With<std::vector<Self>>, With<FieldDescription>, With<KickPlan>, With<Sensors>>().then(
+            on<Trigger<std::vector<VisionBall>>>().then([this](const std::vector<VisionBall>& balls) {
+                if (!balls.empty()) {
+                    ballLastSeen = NUClear::clock::now();
+                }
+            });
+
+            on<Trigger<Ball>, With<Field>, With<FieldDescription>, With<KickPlan>, With<Sensors>>().then(
                 [this](const Ball& ball,
-                       const std::vector<Self>& selfs,
+                       const Field& field,
                        const FieldDescription& fd,
                        const KickPlan& kickPlan,
                        const Sensors& sensors) {
 
                     // Get time since last seen ball
                     auto now = NUClear::clock::now();
                     double secondsSinceLastSeen =
-                        std::chrono::duration_cast<std::chrono::microseconds>(now - ball_last_measurement_time).count()
-                        * 1e-6;
+                        std::chrono::duration_cast<std::chrono::microseconds>(now - ballLastSeen).count() * 1e-6;
 
                     // Compute target in robot coords
-                    auto self = selfs[0];
                     // arma::vec2 kickTarget = {1,0,0}; //Kick forwards
                     // TODO: The heading seems to judder here!!
                     // TODO: use sensors.world instead
-                    arma::vec2 kickTarget = WorldToRobotTransform(convert<double, 2>(self.locObject.position),
-                                                                  convert<double, 2>(self.heading),
-                                                                  convert<double, 2>(kickPlan.target));
-
-                    Transform3D Htw = convert<double, 4, 4>(sensors.world);
-                    arma::vec3 ballPosition =
-                        Htw.transformPoint({ball.locObject.position[0], ball.locObject.position[1], fd.ball_radius});
-                    ball_last_measurement_time = ball.locObject.last_measurement_time;
+                    arma::vec2 kickTarget =
+                        Rotation2D::createRotation(-field.position[2])
+                        * convert<double, 2>(kickPlan.target - Eigen::Vector2d(field.position[0], field.position[1]));
```

Comment:
make a utility for this function OR store a transform 2d in the message. otherwise everyone will replicate this code and make mistakes